Introduction to Graph Neural Networks (#214)
creatis-myriad/creatis-myriad.github.io: collections/_tutorials/2025-03-28-tutorial-graph-neural-networks.md

  or other combinatorial methods applied to the node features.
- Using scoring functions to estimate the likelihood or strength of the connection between nodes.

<div style="text-align:center">
<img src="/collections/images/gnn/edge_level_class.jpg" width=800></div>
<p style="text-align: center;font-style:italic">Figure 9. Example of an edge-level pipeline.</p>

&nbsp;

#### Graph-level tasks

These tasks involve predicting a property or label for the entire graph. To do this, node representations are 
typically aggregated using a global pooling operation to produce an overall graph representation.
The graph prediction $$\hat{y_G}$$ can be formulated as:

$$
h_G = \text{readout}(H^{(L)})
\\
\hat{y_G} = f_{\text{graph}}(h_G)
$$

**Examples:**
- *Computational chemistry:* Predicting molecular properties (e.g., chemical activity) based on the molecular 
  graph structure.
- *Social network analysis:* Classifying an entire network to identify its type or activity level.

**Methods:**
- Multi-layer message passing to refine local node representations by incorporating contextual information from 

>   their neighborhoods. Finally, apply a readout function to aggregate node features into a single graph representation.

Reviewer: Je n'avais pas remarqué cette phrase, qui est redondante avec le bullet point que je t'ai fait ajouté après:
```suggestion
  their neighborhoods.
```

 An MLP head can similarly classify each edge by mapping $$e_{ij}$$ to a probability distribution over edge classes.

&nbsp;

## **Graph Deep Learning paradigm**

Message passing[^13] [^12] is a building block of many popular GNN architectures.
Like convolutions for deep vision models, it is a general framework to process and extract features from the data.
Typical [graph encoders](#encoder-architectures) are designed by stacking multiple message passing steps one after 
the other, to refine features representation on each node (and sometimes edge) in the graph.

### Message passing intuition

Every node $$i$$ sends a message to each of its neighbors $$j$$ and, in turn, receives messages from those neighbors. 
This mechanism enables local information to be disseminated throughout the graph, allowing each node to incorporate 
contextual information from its surroundings. The aggregation process has to ensure permutation invariance, 
since the ordering of neighboring nodes is arbitrary.

&nbsp;

<div style="text-align:center">
<img src="/collections/images/gnn/message_passing.jpg" width=1100></div>

> <p style="text-align: center;font-style:italic">Figure 12. Simple example of message passing (from Sanchez-Lengeling et al.).</p>

Reviewer: On utiliserait plus ce format de citation si tu avais repris la figure telle quelle de l'article Distill (comme je te suggérais au départ). Mais tu l'as suffisamment adapté pour ne pas avoir à citer, puisque ce n'est quand même pas Sanchez-Lengeling qui a inventé le concept de message passing :wink: 
Si tu veux être charitable, tu peux quand même les mentionner, mais avec une formulation un peu plus précise:
```suggestion
<p style="text-align: center;font-style:italic">Figure 12. Simple example of message passing (inspired by a figure from Sanchez-Lengeling et al.[^4]).</p>
```

&nbsp;

#### Classification head

After obtaining the final graph embedding $$h_G$$ through a chosen readout function, the next step is to perform 
classification. As shown in the figure below, we classically use a multi-layer perceptron (MLP) to transform
the embedding $$h_G \in \mathbb{R}^{d_{\text{readout}}}$$ into a logit vector, which is then converted into a 
probability distribution over the target classes.

<div style="text-align:center">
<img src="/collections/images/gnn/classif_head.jpg" width=700></div>
<p style="text-align: center;font-style:italic">Figure 11. Classification head.</p>

&nbsp;

> Note: While $$h_G$$ is used here for graph-level classification, this principle can be extended to other tasks.

- *Node-level classification:* Instead of a single graph embedding, each node $$v_i$$ has an embedding $$h_i$$. 
 Each node embedding is passed individually into the same MLP head to obtain per-node class predictions.
- *Edge-level classification:* Each edge $$(v_i, v_j)$$ has an embedding $$e_{ij}$$. 
 An MLP head can similarly classify each edge by mapping $$e_{ij}$$ to a probability distribution over edge classes.

&nbsp;

## **Graph Deep Learning paradigm**

> Message passing[^13] [^12] lies at the heart of GNN architectures, this step is computed by the 
> [graph encoder](#encoder-architectures), enabling to generate a latent graph on which each node (and sometimes edge) 
> owns an embedded features representation.

Reviewer: Je pensais qu'ici il faut bien contextualiser le rôle du message passing, mais dans un langage accessible pour des gens qui ne sont pas déjà familiers avec le sujet. Qqch comme:
```suggestion
Message passing[^13] [^12] is a building block of many popular GNN architectures.
Like convolutions for deep vision models, it is a general framework to process and extract features from the data.
Typical [graph encoders](#encoder-architectures) are designed by stacking multiple message passing steps one after the other, to refine features representation on each node (and sometimes edge) in the graph.
```

Author: Ok je vois, j'aime bien la comparaison avec la convolution donc je pense que c'est pas mal de garder ta proposition pour présenter le principe :)

$$

$$
E^{(l)} = \begin{bmatrix} e_{12}^{(l)} \\ e_{23}^{(l)} \\ \vdots \\ e_{N-1,N}^{(l)} 
\end{bmatrix} \in \mathbb{R}^{N \times d_e^{(l)}}
\quad \text{with} \quad E = E^{(0)}
$$

&nbsp;

Also, as with the rest of ML models, we denote respectively $$Y$$ and $$\hat{Y}$$ the prediction target and 
the predicted output.

&nbsp;

## **GNN task types**

### Graph-level vs Node-level vs Edge-level tasks

In Graph Neural Networks, tasks are generally categorized based on the scale at which predictions are made. 
These tasks fall into three main categories[^11]:

&nbsp;

#### Graph-level tasks

These tasks involve predicting a property or label for the entire graph. To do this, node representations are 
typically aggregated using a global pooling operation to produce an overall graph representation.
The graph prediction $$\hat{y_G}$$ can be formulated as:

$$
h_G = \text{readout}(H^{(L)})
\\
\hat{y_G} = f_{\text{graph}}(h_G)
$$

**Examples:**
- *Computational chemistry:* Predicting molecular properties (e.g., chemical activity) based on the molecular 
  graph structure.
- *Social network analysis:* Classifying an entire network to identify its type or activity level.

> **Methods:**

Author: J'ai rajouté un **Pooling methods** pour que ce soit plus clair

- Global pooling techniques called [readout functions](#focus-on-graph-level-classification) such as summing, averaging, 
  or taking the maximum of node features, or using attention-based pooling.

> - Readout networks that transform local node representations into a comprehensive global graph representation.

Author: J'avais vu des méthodes plus complexes que de simples fonctions et j'ai peut-être fait l'erreur de trop vite généraliser ça comme des "readout networks". Je vais rajouter des exemples d'articles que j'avais vu passer avec leurs références respectives. Je te laisse me redire ce que t'en penses ;)

- *Edge-level classification:* Each edge $$(v_i, v_j)$$ has an embedding $$e_{ij}$$. 
 An MLP head can similarly classify each edge by mapping $$e_{ij}$$ to a probability distribution over edge classes.

&nbsp;

## **Graph Deep Learning paradigm**

Message passing[^13] [^12] lies at the heart of GNN architectures, this step is computed by the 
[graph encoder](#encoder-architectures), enabling to generate a latent graph on which each node (and sometimes edge) 
owns an embedded features representation.
The main idea is to employ iterative mechanisms, primarily message passing and permutation-invariant 
aggregation, to learn robust representations of graph elements.

### Message passing mechanism

#### Intuition

Every node $$i$$ sends a message to each of its neighbors $$j$$ and, in turn, receives messages from those neighbors. 
This mechanism enables local information to be disseminated throughout the graph, allowing each node to incorporate 
contextual information from its surroundings. The aggregation process also ensures permutation invariance, 
which is essential when dealing with graphs where the ordering of neighboring nodes is arbitrary.

<div style="text-align:center">

> <img src="/collections/images/gnn/forward_prop.jpg" width=650></div>

Author: J'avoue que je ne suis pas très fan de l'illustration de Distill, je trouve qu'elle sous-entend que les informations sont nécessairement sommées puis passer dans une fonction. Ce qui à mon sens est un peu réducteur et fait perdre en abstraction par rapport au principe de base de message + aggregate + update. Je vais rajouter une figure dans ce style dans la section qui suit (pour faire sens avec la formulation mathématique). Je te laisse me redire ce que t'en penses

> Note: $$\mathcal{N}_i$$ can sometimes includes $$i$$ itself, but it is not the convention used in this tutorial.

&nbsp;

### Permutation invariance and neighborhood aggregation

A key challenge in processing graph data is that the neighbors of a node are unordered. 
To ensure that the node representations are independent of the order in which neighbor messages are received, 
the aggregation function must be permutation invariant.

For a node $$v_i$$, the aggregated message is computed as:

$$
m_i^{(l)} = \text{AGGREGATE}\Bigl( \bigl\{ \text{MESSAGE}\bigl( h_j^{(l)}, e_{ij}^{(l)} \bigr) \mid j \in 
\mathcal{N}_i \bigr\} \Bigr)
\\
$$

Since functions like summation, averaging, or maximum are commutative and associative, they guarantee that:

$$
\text{AGGREGATE}\left( \{a, b, c\} \right) = \text{AGGREGATE}\left( \{b, c, a\} \right) = \dots
\\
$$

The node update then becomes:

$$
h_i^{(l+1)} = \text{UPDATE}\bigl( h_i^{(l)}, m_i^{(l)} \bigr)
\\
$$

This design ensures that the final node embedding is invariant to any permutation of its neighbors, 
thereby faithfully reflecting the true structure of the graph.

&nbsp;

> 

Author: Je vais en profiter pour fusionner cette partie avec la précédente du coup, c'est un peu redondant en effet 😅

So the GIN architecture propose to use a combination of the $$\text{Sum}$$ function and an MLP
ensuring a message passing mechanism as expressive as the WL test.

Also, the *Figure 20* below from the original paper[^6] illustrates three cases in which the $$\text{Sum}$$ 
passes the test of injectivity, but not the others.
For each case, three isomorphic graphs are represented, and each node color corresponds to a unique value. 
If the function produces the same output for two or more graphs, then the function failed to distinguish 
isomorphic graphs. The functions are only computed on the neighbors of the center node (the empty one).

&nbsp;

<div style="text-align:center">
<img src="/collections/images/gnn/injectivity_test.jpg" width=700></div>
<p style="text-align: center;font-style:italic">Figure 20. Sum, Mean and Max injectivity comparison.</p>

&nbsp;

This result doesn't mean that the $$\text{Sum}$$ is fully injective (it is not), but it allows us to keep more 
information about the structure of graphs, and thus present a better expressiveness. So, this property is 
particularly important to ensure the [readout function](#focus-on-graph-level-classification) to keep the 

> structure information from the GIN layers by using the $$\text{Sum}$$ function.

Reviewer: Ok! Je comprends mieux avec ton explication. Dans ce cas, je proposerais plus qqch du genre:
> It is particularly beneficial to enforce this injectiveness property in the readout function, since it operates on the full graph, and lots of structural information could be lost with poorly injective readout functions.
